### PR TITLE
python3Packages.pytorch: add pybind11 to buildInputs

### PR DESCRIPTION
**Pull request by jonringer**
###### Description of changes

This allows for the headers to be available, so that
pytorch doesn't try and install pybind11 into its own
includes directory.

 https://www.reddit.com/r/NixOS/comments/v3655n/collision_trying_to_write_my_first_flake/

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by bjornfor — APPROVED** [review]

**Review by robertodr — APPROVED** [review]

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 175962` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.pytorch-metric-learning</li>
    <li>python310Packages.pywick</li>
    <li>python310Packages.shap</li>
    <li>python310Packages.skorch</li>
    <li>python39Packages.pytorch-metric-learning</li>
    <li>python39Packages.pywick</li>
    <li>python39Packages.shap</li>
    <li>python39Packages.skorch</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python310Packages.pymc</li>
    <li>python39Packages.pymc</li>
  </ul>
</details>
<details>
  <summary>62 packages built:</summary>
  <ul>
    <li>openai (python39Packages.openai)</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.boxx</li>
    <li>python310Packages.bpycv</li>
    <li>python310Packages.ffcv</li>
    <li>python310Packages.functorch</li>
    <li>python310Packages.ignite</li>
    <li>python310Packages.openai</li>
    <li>python310Packages.pyro-ppl</li>
    <li>python310Packages.pytorch (python310Packages.pytorchWithoutCuda)</li>
    <li>python310Packages.pytorch-lightning</li>
    <li>python310Packages.pytorch-pfn-extras</li>
    <li>python310Packages.pytorchWithCuda</li>
    <li>python310Packages.qiskit</li>
    <li>python310Packages.qiskit-machine-learning</li>
    <li>python310Packages.rising</li>
    <li>python310Packages.slicer</li>
    <li>python310Packages.spacy</li>
    <li>python310Packages.spacy-loggers</li>
    <li>python310Packages.spacy-transformers</li>
    <li>python310Packages.stanza</li>
    <li>python310Packages.tensorboardx</li>
    <li>python310Packages.test-tube</li>
    <li>python310Packages.textacy</li>
    <li>python310Packages.torch-tb-profiler</li>
    <li>python310Packages.torchgpipe</li>
    <li>python310Packages.torchinfo</li>
    <li>python310Packages.torchmetrics</li>
    <li>python310Packages.torchvision</li>
    <li>python310Packages.wandb</li>
    <li>python310Packages.zcs</li>
    <li>python39Packages.arviz</li>
    <li>python39Packages.boxx</li>
    <li>python39Packages.bpycv</li>
    <li>python39Packages.dalle-mini</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.ffcv</li>
    <li>python39Packages.functorch</li>
    <li>python39Packages.ignite</li>
    <li>python39Packages.pyro-ppl</li>
    <li>python39Packages.pytorch (python39Packages.pytorchWithoutCuda)</li>
    <li>python39Packages.pytorch-lightning</li>
    <li>python39Packages.pytorch-pfn-extras</li>
    <li>python39Packages.pytorchWithCuda</li>
    <li>python39Packages.qiskit</li>
    <li>python39Packages.qiskit-machine-learning</li>
    <li>python39Packages.rising</li>
    <li>python39Packages.slicer</li>
    <li>python39Packages.spacy</li>
    <li>python39Packages.spacy-loggers</li>
    <li>python39Packages.spacy-transformers</li>
    <li>python39Packages.stanza</li>
    <li>python39Packages.tensorboardx</li>
    <li>python39Packages.test-tube</li>
    <li>python39Packages.textacy</li>
    <li>python39Packages.torch-tb-profiler</li>
    <li>python39Packages.torchgpipe</li>
    <li>python39Packages.torchinfo</li>
    <li>python39Packages.torchmetrics</li>
    <li>python39Packages.torchvision</li>
    <li>python39Packages.wandb</li>
    <li>python39Packages.zcs</li>
  </ul>
</details>

**Comment by jonringer** [comment]
the pymc failures are not related

```
https://github.com/NixOS/nixpkgs/pull/175962

8 packages marked as broken and skipped:
python310Packages.pytorch-metric-learning python310Packages.pywick python310Packages.shap python310Packages.skorch python39Packages.pytorch-metric-learning python39Packages.pywick python39Packages.shap python39Packages.skorch

2 packages failed to build:
python310Packages.pymc python39Packages.pymc

62 packages built:
openai python310Packages.arviz python310Packages.boxx python310Packages.bpycv python310Packages.ffcv python310Packages.functorch python310Packages.ignite python310Packages.openai python310Packages.pyro-ppl python310Packages.pytorch python310Packages.pytorch-lightning python310Packages.pytorch-pfn-extras python310Packages.pytorchWithCuda python310Packages.qiskit python310Packages.qiskit-machine-learning python310Packages.rising python310Packages.slicer python310Packages.spacy python310Packages.spacy-loggers python310Packages.spacy-transformers python310Packages.stanza python310Packages.tensorboardx python310Packages.test-tube python310Packages.textacy python310Packages.torch-tb-profiler python310Packages.torchgpipe python310Packages.torchinfo python310Packages.torchmetrics python310Packages.torchvision python310Packages.wandb python310Packages.zcs python39Packages.arviz python39Packages.boxx python39Packages.bpycv python39Packages.dalle-mini python39Packages.elegy python39Packages.ffcv python39Packages.functorch python39Packages.ignite python39Packages.pyro-ppl python39Packages.pytorch python39Packages.pytorch-lightning python39Packages.pytorch-pfn-extras python39Packages.pytorchWithCuda python39Packages.qiskit python39Packages.qiskit-machine-learning python39Packages.rising python39Packages.slicer python39Packages.spacy python39Packages.spacy-loggers python39Packages.spacy-transformers python39Packages.stanza python39Packages.tensorboardx python39Packages.test-tube python39Packages.textacy python39Packages.torch-tb-profiler python39Packages.torchgpipe python39Packages.torchinfo python39Packages.torchmetrics python39Packages.torchvision python39Packages.wandb python39Packages.zcs
```

**Comment by samuela** [comment]
Yeah, those seem to be caused by
```
error: hash mismatch in fixed-output derivation '/nix/store/m604ai7h3qn0f973drm3bjkpl0d1122k-source.drv':
         specified: sha256-NJxklOpIbSbi/SB/rafBNllpnNb1yWLVpyB2f/U0i78=
            got:    sha256-vIDyqYj9HuH5gVcc6KPkB2NNTAggiPHyAR2fCXOQzsE=
error: 1 dependencies of derivation '/nix/store/whd26jmzyy9pp21pf2ar0zm89v068f9g-python3.10-aesara-2.7.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ix2a43rb86bqpmmy3wvx2kf12lgj72fx-python3.9-aesara-2.7.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/h0cmjf5ai4xjicw65injlf9aa62zibpz-python3.10-aeppl-0.0.31.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7n45ig3jsyh57l09zdrd2y4scigh80p6-python3.9-aeppl-0.0.31.drv' failed to build
error: 2 dependencies of derivation '/nix/store/241ym0x696c5djb6g32j5w3ss2d0i369-python3.10-pymc-4.0.0.drv' failed to build
error: 2 dependencies of derivation '/nix/store/mghlxad1xcwli8jjpk16i5al351r39gl-python3.9-pymc-4.0.0.drv' failed to build
```

File: pkgs/development/python-modules/functorch/default.nix [hunk 1]
```diff
@@ -29,6 +30,10 @@ buildPythonPackage rec {
     which
   ];
 
+  buildInputs = [
+    pybind11
+  ];
```
Comment (samuela) [inline pkgs/development/python-modules/functorch/default.nix:35]: I'm assuming this proved necessary in order to get the build to go through? I guess it was relying on PyTorch's pybind11 headers previously?

Comment (jonringer) [reply pkgs/development/python-modules/functorch/default.nix:35]: correct, it was able to find pybind headers through pytorch

This seems to be the only package which had this behavior

**Review by samuela — APPROVED** [review]

**Pull request merged**